packages/core/src/jsx.ts

```diff
@@ -0,0 +1,121 @@
+import { capitalize, memo } from '@pandacss/shared'
+import type { Context } from './context'
+import type { PatternNode } from './patterns'
+import type { RecipeNode } from './types'
+
+interface JsxMatcher {
+  string: Set<string>
+  regex: RegExp[]
+}
+
+export class JsxEngine {
+  nodes: Array<PatternNode | RecipeNode> = []
+
+  recipeMatcher: JsxMatcher = { string: new Set(), regex: [] }
+  recipePropertiesByJsxName = new Map<string, Set<string>>()
+
+  patternMatcher: JsxMatcher = { string: new Set(), regex: [] }
+  patternPropertiesByJsxName = new Map<string, Set<string>>()
+
+  constructor(private context: Pick<Context, 'patterns' | 'recipes' | 'config'>) {
+    this.nodes = [...context.patterns.details, ...context.recipes.details]
+    this.assignRecipeMatcher()
+    this.assignPatternMatcher()
+  }
+
+  assignRecipeMatcher() {
+    if (!this.isEnabled) return
+
+    for (const recipe of this.context.recipes.details) {
+      this.recipePropertiesByJsxName.set(recipe.jsxName, new Set(recipe.props ?? []))
+      recipe.jsx.forEach((jsx) => {
+        if (typeof jsx === 'string') {
+          this.recipeMatcher.string.add(jsx)
+        } else {
+          this.recipeMatcher.regex.push(jsx)
+        }
+      })
+    }
+  }
+
+  assignPatternMatcher() {
+    if (!this.isEnabled) return
+
+    for (const pattern of this.context.patterns.details) {
+      this.patternPropertiesByJsxName.set(pattern.jsxName, new Set(pattern.props ?? []))
+      pattern.jsx.forEach((jsx) => {
+        if (typeof jsx === 'string') {
+          this.patternMatcher.string.add(jsx)
+        } else {
+          this.patternMatcher.regex.push(jsx)
+        }
+      })
+    }
+  }
+
+  get names() {
+    return [this.factoryName, ...this.nodes.map((node) => node.jsxName)]
+  }
```

Comment:
this will get re-executed on each property access, we probably want to save it to a property class instead since it loops on each nodes